File: R/nprobust/tests/testthat/test-weights.R
## User weights should multiply kernel weights throughout: point estimates
## must match a weighted least squares fit with w = K(u) * user_weights.

test_that("weights arg matches WLS with kernel * user weights", {
  fx <- make_fixture(n = 600)
  set.seed(3)
  w <- runif(fx$n, 0.2, 2)

  h <- 0.3; eval <- 0; p <- 1
  est <- lprobust(fx$y, fx$x, eval = eval, h = h, p = p,
                  kernel = "uni", vce = "hc0", weights = w)

  u <- (fx$x - eval) / h
  kw <- w_fun(u, "uni")
  comb <- kw * w
  sub  <- comb > 0
  xc <- fx$x[sub] - eval
  ys <- fx$y[sub]
  ws <- comb[sub]
  X <- cbind(1, xc)
  beta <- solve(crossprod(X * sqrt(ws)), crossprod(X * ws, ys))

  expect_equal(as.numeric(est$Estimate[, "tau.us"]),
               as.numeric(beta[1]),
               tolerance = 1e-10)
})

test_that("weights=1 matches no weights", {
  fx <- make_fixture(n = 400)
  est0 <- lprobust(fx$y, fx$x, eval = 0, h = 0.3, p = 1, kernel = "epa", vce = "hc0")
  est1 <- lprobust(fx$y, fx$x, eval = 0, h = 0.3, p = 1, kernel = "epa", vce = "hc0",
                   weights = rep(1, fx$n))
  expect_equal(as.numeric(est0$Estimate), as.numeric(est1$Estimate),
               tolerance = 1e-12)
})

test_that("zero weights exclude observations", {
  fx <- make_fixture(n = 400)
  w <- rep(1, fx$n); w[fx$x > 0.5] <- 0
  est_w <- lprobust(fx$y, fx$x, eval = 0.3, h = 0.3, p = 1, kernel = "uni",
                    vce = "hc0", weights = w)
  est_s <- lprobust(fx$y, fx$x, eval = 0.3, h = 0.3, p = 1, kernel = "uni",
                    vce = "hc0", subset = fx$x <= 0.5)
  expect_equal(as.numeric(est_w$Estimate[, "tau.us"]),
               as.numeric(est_s$Estimate[, "tau.us"]),
               tolerance = 1e-12)
})

File: Python/nprobust/tests/test_weights.py
import numpy as np
import pytest

from nprobust import lprobust
from .conftest import w_kernel


def test_weights_equal_wls(fixture_data):
    x, y = fixture_data["x"], fixture_data["y"]
    rng = np.random.default_rng(3)
    w = rng.uniform(0.2, 2, size=len(x))
    h = 0.3
    m = lprobust(y, x, eval=np.array([0.0]), h=h, p=1, kernel="uni",
                 vce="hc0", weights=w)

    u = (x - 0.0) / h
    kw = w_kernel(u, "uni")
    comb = kw * w
    sub = comb > 0
    xc = x[sub]; ys = y[sub]; ws = comb[sub]
    X = np.column_stack([np.ones_like(xc), xc])
    Xs = X * np.sqrt(ws)[:, None]
    beta = np.linalg.lstsq(Xs, ys * np.sqrt(ws), rcond=None)[0]
    assert m.Estimate["tau.us"].iloc[0] == pytest.approx(beta[0], abs=1e-10)


def test_weights_ones_matches_unweighted(fixture_data):
    x, y = fixture_data["x"], fixture_data["y"][:400]
    x = x[:400]
    m0 = lprobust(y, x, eval=np.array([0.0]), h=0.3, p=1, kernel="epa", vce="hc0")
    m1 = lprobust(y, x, eval=np.array([0.0]), h=0.3, p=1, kernel="epa", vce="hc0",
                  weights=np.ones_like(x))
    assert m0.Estimate["tau.us"].iloc[0] == pytest.approx(
        m1.Estimate["tau.us"].iloc[0], abs=1e-12)


def test_zero_weights_match_subset(fixture_data):
    x, y = fixture_data["x"][:400], fixture_data["y"][:400]
    w = np.ones_like(x); w[x > 0.5] = 0
    mw = lprobust(y, x, eval=np.array([0.3]), h=0.3, p=1, kernel="uni",
                  vce="hc0", weights=w)
    ms = lprobust(y, x, eval=np.array([0.3]), h=0.3, p=1, kernel="uni",
                  vce="hc0", subset=(x <= 0.5))
    assert mw.Estimate["tau.us"].iloc[0] == pytest.approx(
        ms.Estimate["tau.us"].iloc[0], abs=1e-12)
